module/zfs/bpobj.c

```diff
@@ -206,155 +206,213 @@ bpobj_is_empty(bpobj_t *bpo)
+				ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
+				if (bpo->bpo_havecomp) {
+					bpo->bpo_phys->bpo_comp -=
+					    BP_GET_PSIZE(bp);
+					bpo->bpo_phys->bpo_uncomp -=
+					    BP_GET_UCSIZE(bp);
+					comp_freed += BP_GET_PSIZE(bp);
+					uncomp_freed += BP_GET_UCSIZE(bp);
+				}
+				bpo->bpo_phys->bpo_num_blkptrs--;
+				ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs,
+				    >=, 0);
+			}
+		}
+		if (dbuf) {
+			dmu_buf_rele(dbuf, FTAG);
+			dbuf = NULL;
 		}
-
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
-
-		bparray = dbuf->db_data;
-		bp = &bparray[blkoff];
-		err = func(arg, bp, tx);
-		if (err)
-			break;
 		if (free) {
-			bpo->bpo_phys->bpo_bytes -=
-			    bp_get_dsize_sync(dmu_objset_spa(bpo->bpo_os), bp);
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			if (bpo->bpo_havecomp) {
-				bpo->bpo_phys->bpo_comp -= BP_GET_PSIZE(bp);
-				bpo->bpo_phys->bpo_uncomp -= BP_GET_UCSIZE(bp);
+			VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os,
+			    bpo->bpo_object, (i + 1) * sizeof (blkptr_t),
+			    DMU_OBJECT_END, tx));
+		}
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				if (free) {
+					bpobj_phys_t *bpop =
+					    bpi->bpi_parent->bpo_phys;
+					bpop->bpo_bytes -= freed;
+					ASSERT3S(bpop->bpo_bytes, >=, 0);
+					bpop->bpo_comp -= comp_freed;
+					bpop->bpo_uncomp -= uncomp_freed;
+				}
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
 			}
-			bpo->bpo_phys->bpo_num_blkptrs--;
-			ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs, >=, 0);
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			goto out;
 		}
-	}
-	if (dbuf) {
-		dmu_buf_rele(dbuf, FTAG);
-		dbuf = NULL;
-	}
-	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
-	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
-
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
-	}
-	ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
-	epb = doi.doi_data_block_size / sizeof (uint64_t);
 
-	for (i = bpo->bpo_phys->bpo_num_subobjs - 1; i >= 0; i--) {
-		uint64_t *objarray;
-		uint64_t offset, blkoff;
-		bpobj_t sublist;
-		uint64_t used_before, comp_before, uncomp_before;
-		uint64_t used_after, comp_after, uncomp_after;
+		if (!bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0) {
+			/* propogate reduction in space used and clean up */
+			if (free && (bpi->bpi_parent != NULL)) {
+				bpobj_t *p = bpi->bpi_parent;
+				err = dmu_object_free(p->bpo_os,
+				    bpi->bpi_object, tx);
+				if (err)
+					break;
+				p->bpo_phys->bpo_num_subobjs--;
+				ASSERT3S(p->bpo_phys->bpo_num_subobjs, >=, 0);
+				VERIFY3U(0, ==, dmu_free_range(p->bpo_os,
+				    p->bpo_phys->bpo_subobjs,
+				    bpi->bpi_index * sizeof (uint64_t),
+				    DMU_OBJECT_END, tx));
```

Comment:
Are we sure that, given the order we are processing these in, that this is only chopping off a single entry?  Could we somehow verify that?  e.g. by checking that `bpi_index == bpo_num_subobjs-1`

Reply:
I have added an ASSERT for this.
```
                if (!bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0) {
                        /* propogate reduction in space used and clean up */
                        if (free && (bpi->bpi_parent != NULL)) {
                                bpobj_t *p = bpi->bpi_parent;
                                err = dmu_object_free(p->bpo_os,
                                    bpi->bpi_object, tx);
                                if (err)
                                        break;
                                ASSERT3S(bpi->bpi_index, ==, p->bpo_phys->bpo_num_subobjs - 1);
                                cmn_err(CE_NOTE, "clearing range at bpi_index %llu num_subojbs %llu",
                                    (u_longlong_t)bpi->bpi_index, (u_longlong_t)p->bpo_phys->bpo_num_subobjs);
                                p->bpo_phys->bpo_num_subobjs--;
                                ASSERT3S(p->bpo_phys->bpo_num_subobjs, >=, 0);
                                VERIFY3U(0, ==, dmu_free_range(p->bpo_os,
                                    p->bpo_phys->bpo_subobjs,
                                    bpi->bpi_index * sizeof (uint64_t),
                                    DMU_OBJECT_END, tx));
                                p->bpo_phys->bpo_bytes -= freed;
                                ASSERT3S(p->bpo_phys->bpo_bytes, >=, 0);
                                p->bpo_phys->bpo_comp -= comp_freed;
                                p->bpo_phys->bpo_uncomp -= uncomp_freed;
                        }
```
And here is the debug output:
```
[ 1713.047827] NOTICE: clearing range at bpi_index 0 num_subojbs 1
[ 1713.210079] NOTICE: clearing range at bpi_index 34 num_subojbs 35
[ 1713.248953] NOTICE: clearing range at bpi_index 33 num_subojbs 34
[ 1713.282443] NOTICE: clearing range at bpi_index 32 num_subojbs 33
[ 1713.315946] NOTICE: clearing range at bpi_index 31 num_subojbs 32
[ 1713.349813] NOTICE: clearing range at bpi_index 30 num_subojbs 31
[ 1713.383331] NOTICE: clearing range at bpi_index 29 num_subojbs 30
[ 1713.418360] NOTICE: clearing range at bpi_index 28 num_subojbs 29
[ 1713.442221] NOTICE: clearing range at bpi_index 27 num_subojbs 28
[ 1713.479142] NOTICE: clearing range at bpi_index 26 num_subojbs 27
[ 1713.511587] NOTICE: clearing range at bpi_index 25 num_subojbs 26
[ 1713.543077] NOTICE: clearing range at bpi_index 24 num_subojbs 25
[ 1713.614529] NOTICE: clearing range at bpi_index 23 num_subojbs 24
[ 1713.685240] NOTICE: clearing range at bpi_index 22 num_subojbs 23
[ 1713.743492] NOTICE: clearing range at bpi_index 21 num_subojbs 22
[ 1713.802604] NOTICE: clearing range at bpi_index 20 num_subojbs 21
[ 1713.869107] NOTICE: clearing range at bpi_index 19 num_subojbs 20
[ 1713.918777] NOTICE: clearing range at bpi_index 18 num_subojbs 19
[ 1713.973525] NOTICE: clearing range at bpi_index 17 num_subojbs 18
[ 1714.038343] NOTICE: clearing range at bpi_index 16 num_subojbs 17
[ 1714.155260] NOTICE: clearing range at bpi_index 15 num_subojbs 16
[ 1714.591379] NOTICE: clearing range at bpi_index 14 num_subojbs 15
[ 1714.622209] NOTICE: clearing range at bpi_index 13 num_subojbs 14
[ 1714.655231] NOTICE: clearing range at bpi_index 12 num_subojbs 13
[ 1714.688806] NOTICE: clearing range at bpi_index 11 num_subojbs 12
[ 1714.716590] NOTICE: clearing range at bpi_index 10 num_subojbs 11
[ 1714.752010] NOTICE: clearing range at bpi_index 9 num_subojbs 10
[ 1714.790763] NOTICE: clearing range at bpi_index 8 num_subojbs 9
[ 1714.826907] NOTICE: clearing range at bpi_index 7 num_subojbs 8
[ 1714.863624] NOTICE: clearing range at bpi_index 6 num_subojbs 7
[ 1714.903465] NOTICE: clearing range at bpi_index 5 num_subojbs 6
[ 1714.937283] NOTICE: clearing range at bpi_index 4 num_subojbs 5
[ 1714.969658] NOTICE: clearing range at bpi_index 3 num_subojbs 4
[ 1715.004930] NOTICE: clearing range at bpi_index 2 num_subojbs 3
[ 1715.035054] NOTICE: clearing range at bpi_index 1 num_subojbs 2
[ 1715.069743] NOTICE: clearing range at bpi_index 0 num_subojbs 1
```